zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1088 @@
+        vec![
+            ScanRange::from_parts(extended.start..range.start, ScanPriority::FoundNote),
+            ScanRange::from_parts(range.end..extended.end, ScanPriority::FoundNote),
+        ]
+    } else {
+        vec![]
+    };
+
+    replace_queue_entries(
+        conn,
+        &query_range,
+        Some(scanned).into_iter().chain(extensions.into_iter()),
+    )?;
+
+    Ok(())
+}
+
+pub(crate) fn update_chain_tip<P: consensus::Parameters>(
+    conn: &rusqlite::Transaction<'_>,
+    params: &P,
+    new_tip: BlockHeight,
+) -> Result<(), SqliteClientError> {
+    // `ScanRange` uses an exclusive upper bound.
+    let chain_end = new_tip + 1;
+
+    // Read the maximum height from the shards table.
+    let shard_start_height = conn.query_row(
+        "SELECT MAX(subtree_end_height)
+        FROM sapling_tree_shards",
+        [],
+        |row| Ok(row.get::<_, Option<u32>>(0)?.map(BlockHeight::from)),
+    )?;
+
+    // Create a scanning range for the fragment of the last shard leading up to new tip.
+    let shard_entry = shard_start_height
+        .filter(|h| h < &chain_end)
+        .map(|h| ScanRange::from_parts(h..chain_end, ScanPriority::ChainTip));
+
+    // Create scanning ranges to either validate potentially invalid blocks at the wallet's view
+    // of the chain tip, or connect the prior tip to the new tip.
+    let tip_entry = block_height_extrema(conn)?.map(|(_, prior_tip)| {
+        // If we don't have shard metadata, this means we're doing linear scanning, so create a
+        // scan range from the prior tip to the current tip with `Historic` priority.
+        if shard_entry.is_none() {
+            ScanRange::from_parts(prior_tip..chain_end, ScanPriority::Historic)
+        } else {
+            // Determine the height to which we expect blocks retrieved from the block source to be stable
+            // and not subject to being reorg'ed.
+            let stable_height = new_tip.saturating_sub(PRUNING_DEPTH);
+
+            // if the wallet's prior tip is above the stable height, prioritize the range between
+            // it and the new tip as `ChainTip`. Otherwise, prioritize the `VALIDATION_DEPTH`
+            // blocks above the wallet's prior tip as `Verify`. Since `scan_cached_blocks`
+            // retrieves the metadata for the block being connected to, the connectivity to the
+            // prior tip will always be checked. Since `Verify` ranges have maximum priority, even
+            // if the block source begins downloading blocks from the shard scan range (which ends
+            // at the stable height) the scanner should not attempt to scan those blocks until the
+            // tip range has been completely checked and any required rewinds have been performed.
+            if prior_tip >= stable_height {
+                // This may overlap the `shard_entry` range and if so will be coalesced with it.
+                ScanRange::from_parts(prior_tip..chain_end, ScanPriority::ChainTip)
```

Comment:
If `chain_end >= prior_tip` above then that also holds here, but again, how do we know that?

Reply:
Pushed a commit that does nothing in this scenario.